### PR TITLE
Fixes #11230  Problem with parsing of form parameters without values.

### LINKED ISSUE #11230: Problem with parsing of form parameters without values in Jetty 12?
**Jetty version(s)**
12.0.5

**Jetty Environment**
ee10 

**Java version/vendor** `(use: java -version)`
```
openjdk version "21.0.1" 2023-10-17 LTS
OpenJDK Runtime Environment Temurin-21.0.1+12 (build 21.0.1+12-LTS)
OpenJDK 64-Bit Server VM Temurin-21.0.1+12 (build 21.0.1+12-LTS, mixed mode)
```

**OS type/version**
macOS Sonoma 14.2.1

**Description**

POST-ing a form which contains parameters without values causes the `request.getParameterNames()` to return wrong values for parameter names.

**How to reproduce?**

I have this simple servlet running with `jetty-ee10-maven-plugin`
and Jetty 12.0.5:

```java
package test;

import java.io.IOException;
import java.io.PrintWriter;
import java.util.Enumeration;
import jakarta.servlet.ServletException;
import jakarta.servlet.annotation.WebServlet;
import jakarta.servlet.http.HttpServlet;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

@WebServlet(value="/", name="helloWorldServlet")
public class HelloWorldServlet extends HttpServlet {

    @Override
    protected void service(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
        resp.setStatus(200);
        resp.setHeader("Content-Type", "text/plain;charset=utf-8");
        
        try (PrintWriter writer = resp.getWriter()) {
            writer.write("Method: " + req.getMethod() + "\n");
            writer.write("Parameters:\n");
            Enumeration<String> names = req.getParameterNames();
            while (names.hasMoreElements()) {
                String name = names.nextElement();
                writer.write("   " + name + " = '" + req.getParameter(name) + "'\n");
            }
        }
    }
}
```

When I try this with `curl`, I get the following output:
```
> curl -X POST 'http://localhost:8080/?param0=value0&param1' -d 'param2&param3=value3&param4=value4&param5'
Method: POST
Parameters:
   param0 = 'value0'
   param1 = ''
   param2&param3 = 'value3'
   param4 = 'value4'
```

The parameters supplied in the query string (`param0`, `param1`) seem
OK, however there is something strange with the parsing of the form
parameters in the request body (`param2&param3` should be separated,
`param5` is missing).

This is the output when running the same `curl` command  with the same servlet (only modified to use `javax.servlet`) running under Jetty 9:
```
Method: POST
Parameters:
   param5 = ''
   param0 = 'value0'
   param3 = 'value3'
   param4 = 'value4'
   param1 = ''
   param2 = ''
```

### PR DESCRIPTION
**Pull request by sbordet**
Fixed parsing of form parameters in FormFields.parse(). 
Added more tests for valid edge cases, and invalid cases.

### CODE REVIEW

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/FormFieldsTest.java [hunk 1]
```diff
@@ -89,4 +105,32 @@ public void testFormFields(List<String> chunks, Charset charset, int maxFields,
             assertNull(expected);
         }
     }
+
+    public static Stream<Arguments> invalidData()
+    {
+        return Stream.of(
+            Arguments.of(List.of("n=v&X=Y"), UTF_8, 1, -1, IllegalStateException.class),
+            Arguments.of(List.of("n=v&X=Y"), UTF_8, -1, 3, IllegalStateException.class),
```
Comment (joakime) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/FormFieldsTest.java:124]: Why are these 2 invalid?
Also, they look identical on this review, but the maxFields / maxLength is different?

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/FormFieldsTest.java [hunk 2]
```diff
@@ -89,4 +105,32 @@ public void testFormFields(List<String> chunks, Charset charset, int maxFields,
...
+            Arguments.of(List.of("=v"), UTF_8, -1, -1, IllegalStateException.class),
+            Arguments.of(List.of("n%AH=v"), UTF_8, -1, -1, IllegalArgumentException.class),
+            Arguments.of(List.of("n=v%AH"), UTF_8, -1, -1, IllegalArgumentException.class),
+            Arguments.of(List.of("n=v%FF"), UTF_8, -1, -1, CharacterCodingException.class)
+        );
```
Comment (joakime) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/FormFieldsTest.java:128]: Some new test cases ...

* `"n=v%0A"` what happens if a value has a control character?
* `"n%0D=v"` what happens if a name has a control character?
* `"n%00=v"` what happens if a name has a null character?

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/FormFieldsTest.java [hunk 3]
```diff
@@ -32,31 +34,44 @@
...
 {
-    public static Stream<Arguments> tests()
+    public static Stream<Arguments> validData()
     {
         return Stream.of(
+            Arguments.of(List.of(""), UTF_8, -1, -1, Map.of()),
+            Arguments.of(List.of("name"), UTF_8, -1, -1, Map.of("name", "")),
+            Arguments.of(List.of("name&"), UTF_8, -1, -1, Map.of("name", "")),
+            Arguments.of(List.of("name="), UTF_8, -1, -1, Map.of("name", "")),
+            Arguments.of(List.of("%A"), UTF_8, -1, -1, Map.of("%A", "")),
```
Comment (joakime) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/FormFieldsTest.java]: single character pct-encoding?
This seems like a bug, and something we shouldn't support.

**Review by joakime — CHANGES_REQUESTED** [review]
Needs a few more test cases, and some test cases moved to the invalid side.

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/FormFieldsTest.java (hunk 1 above)
Comment (sbordet) [reply jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/FormFieldsTest.java:124]: maxFields=1 for the first, maxLength=3 for the second.

**Review by joakime — APPROVED** [review]

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/FormFieldsTest.java (hunk 2 above)
Comment (sbordet) [reply jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/FormFieldsTest.java:128]: Apparently control chars are allowed.

**Review by lorban — APPROVED** [review]
